components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java

```diff
@@ -1100,6 +1107,112 @@ private void handleRoleClaim(AuthenticationContext context, Map<String, String>
+     *
+     * @param stepConfig StepConfig of current step.
+     * @param context AuthenticationContext of current authentication flow.
+     * @param allLocalClaims All local claims of the current authenticated user.
+     * @throws FrameworkException Exception on handling application roles for local user.
+     */
+    protected void handleApplicationRolesForLocalUser(StepConfig stepConfig, AuthenticationContext context,
+                                                    Map<String, String> allLocalClaims)
+            throws FrameworkException {
+
+        AuthenticatedUser authenticatedUser = (stepConfig == null)
+                ? context.getSequenceConfig().getAuthenticatedUser()
+                : stepConfig.getAuthenticatedUser();
+        if (stepConfig == null || stepConfig.isSubjectAttributeStep()) {
+            String requestedAppRoleClaim = context.getSequenceConfig().getApplicationConfig()
+                    .getRequestedClaimMappings().get(FrameworkConstants.APP_ROLES_CLAIM);
+            if (requestedAppRoleClaim != null) {
+                String appRoles = getApplicationRoles(authenticatedUser, context);
+                if (appRoles != null) {
+                    allLocalClaims.put(FrameworkConstants.APP_ROLES_CLAIM, appRoles);
+                }
+            }
+        }
+    }
+
+    /**
+     * Get the application roles of the authenticated user for application roles resolver if available.
+     *
+     * @param authenticatedUser Authenticated user to get the application roles for.
+     * @param context Authentication context.
+     * @return Application roles of the authenticated user.
+     * @throws FrameworkException Exception on getting application roles.
+     */
+    private String getApplicationRoles(AuthenticatedUser authenticatedUser, AuthenticationContext context)
+            throws FrameworkException {
+
+        String applicationId = context.getSequenceConfig().getApplicationConfig().getServiceProvider()
+                .getApplicationResourceId();
+        ApplicationRolesResolver appRolesResolver = FrameworkServiceDataHolder.getInstance()
+                .getHighestPriorityApplicationRolesResolver();
+        if (appRolesResolver == null) {
+            log.debug("No application roles resolver found.");
+            // Return empty string if no application roles resolver is available.
+            return StringUtils.EMPTY;
+        }
+        String[] appRoles;
+        try {
+            appRoles = appRolesResolver.getRoles(authenticatedUser, applicationId);
+        } catch (ApplicationRolesException e) {
+            throw new FrameworkException("Error while retrieving application roles for user: " +
+                    authenticatedUser + " and application: " + applicationId, e);
+        }
+        if (appRoles != null) {
+            return String.join(FrameworkUtils.getMultiAttributeSeparator(), appRoles);
+        }
+        return null;
```

Comment:
Should this be an empty string ?

Reply:
This can be null because all scenarios where empty string should be returned are handled already.